cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,474 @@
+# DNA: Proximal Policy Optimization with a Dual Network Architecture
+# https://arxiv.org/abs/2206.10027
+
+import argparse
+import os
+import random
+import time
+from collections import deque
+from copy import deepcopy
+from distutils.util import strtobool
+
+import envpool
+import gym
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.distributions import Categorical, kl_divergence
+from torch.utils.tensorboard import SummaryWriter
+
+
+def parse_args():
+    # fmt: off
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--exp-name", type=str, default=os.path.basename(__file__).rstrip(".py"),
+        help="the name of this experiment")
+    parser.add_argument("--seed", type=int, default=np.random.randint(2 ** 31),
+        help="seed of the experiment")
+    parser.add_argument("--torch-deterministic", type=lambda x: bool(strtobool(x)), default=True, nargs="?", const=True,
+        help="if toggled, `torch.backends.cudnn.deterministic=False`")
+    parser.add_argument("--cuda", type=lambda x: bool(strtobool(x)), default=True, nargs="?", const=True,
+        help="if toggled, cuda will be enabled by default")
+    parser.add_argument("--track", type=lambda x: bool(strtobool(x)), default=False, nargs="?", const=True,
+        help="if toggled, this experiment will be tracked with Weights and Biases")
+    parser.add_argument("--wandb-project-name", type=str, default="cleanRL",
+        help="the wandb's project name")
+    parser.add_argument("--wandb-entity", type=str, default=None,
+        help="the entity (team) of wandb's project")
+
+    # Algorithm specific arguments
+    parser.add_argument("--env-id", type=str, default="Pong-v5",
+        help="the id of the environment")
+    parser.add_argument("--total-timesteps", type=int, default=10_000_000,
+        help="total timesteps of the experiments")
+    parser.add_argument("--num-envs", type=int, default=128,
+        help="the number of parallel game environments")
+    parser.add_argument("--envpool-num-threads", type=int, default=2,
+        help="the number of envpool worker threads")
+    parser.add_argument("--num-steps", type=int, default=128,
+        help="rollout horizon: the number of steps to run in each environment per policy rollout")
+    parser.add_argument("--anneal-lr", type=lambda x: bool(strtobool(x)), default=True, nargs="?", const=True,
+        help="Toggle learning rate annealing for policy and value networks")
+    parser.add_argument("--gamma", type=float, default=0.99,
```

Comment:
Note a complete review yet, just putting a note here: the author suggests the `gamma` parameter used in DNA should be 0.999. I am ok with using either 0.99 as long as this difference is spelled out.

Reply:
I'd recommend using the same settings as the paper, which would be the following changes to the defaults

gamma=0.999 
clip_coef=0.2
anneal-lr=false 
max_grad_norm=5.0

Gamma and clip_coef  make a fairly big difference too. The other settings (max_grad_norm, anneal etc) are not so important.

Reply:
yeah, I was originally thinking to have minimal changes from vanilla PPO hyperparams, but I agree it's better to have well tuned algorithm specific defaults out of the box.

Reply:
I'm also updating entropy coef, as it's 0.001 in the paper, vs. 0.01 in baseline PPO